server/sonar-server/src/main/java/org/sonar/server/computation/step/FeedMeasureComputers.java

```diff
@@ -0,0 +1,174 @@
+
+  /**
+   * Constructor override used by Pico to instantiate the class when no plugin is defining metrics
+   */
+  public FeedMeasureComputers(MeasureComputersHolderImpl measureComputersHolder, MeasureComputerProvider[] measureComputerProviders) {
+    this(measureComputersHolder, new Metrics[] {}, measureComputerProviders);
+  }
+
+  /**
+   * Constructor override used by Pico to instantiate the class when no plugin is defining measure computers
+   */
+  public FeedMeasureComputers(MeasureComputersHolderImpl measureComputersHolder, Metrics[] metricsRepositories) {
+    this(measureComputersHolder, metricsRepositories, new MeasureComputerProvider[] {});
+  }
+
+  /**
+   * Constructor override used by Pico to instantiate the class when no plugin is defining metrics neither measure computers
+   */
+  public FeedMeasureComputers(MeasureComputersHolderImpl measureComputersHolder) {
+    this(measureComputersHolder, new Metrics[] {}, new MeasureComputerProvider[] {});
+  }
+
+  @Override
+  public void execute() {
+    MeasureComputerProviderContext context = new MeasureComputerProviderContext();
+    for (MeasureComputerProvider provider : measureComputerProviders) {
+      provider.register(context);
+    }
+    validateInputMetrics(context.getMeasureComputers());
+    measureComputersHolder.setMeasureComputers(sortComputers(context.getMeasureComputers()));
+  }
+
+  private static Iterable<MeasureComputer> sortComputers(List<MeasureComputer> computers) {
+    Map<String, MeasureComputer> computersByOutputMetric = new HashMap<>();
+    Map<String, MeasureComputer> computersByInputMetric = new HashMap<>();
+    for (MeasureComputer computer : computers) {
+      for (String outputMetric : computer.getOutputMetrics()) {
+        computersByOutputMetric.put(outputMetric, computer);
+      }
+      for (String inputMetric : computer.getInputMetrics()) {
+        computersByInputMetric.put(inputMetric, computer);
+      }
+    }
+    ToComputerByKey toComputerByOutputMetricKey = new ToComputerByKey(computersByOutputMetric);
+    ToComputerByKey toComputerByInputMetricKey = new ToComputerByKey(computersByInputMetric);
+
+    DirectAcyclicGraph dag = new DirectAcyclicGraph();
+    for (MeasureComputer computer : computers) {
+      dag.add(computer);
+      for (MeasureComputer dependency : getDependencies(computer, toComputerByOutputMetricKey)) {
+        dag.add(computer, dependency);
+      }
+      for (MeasureComputer generates : getDependents(computer, toComputerByInputMetricKey)) {
+        dag.add(generates, computer);
+      }
+    }
+    return FluentIterable.from(dag.sort());
+  }
+
+  private void validateInputMetrics(List<MeasureComputer> computers) {
+    Set<String> pluginMetricKeys = FluentIterable.from(Arrays.asList(metricsRepositories))
+      .transformAndConcat(MetricsToMetricList.INSTANCE)
+      .transform(MetricToKey.INSTANCE)
+      .toSet();
+    // TODO would be nice to generate an error containing all bad input metrics
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)